compiler/lib/src/main/scala/codegen/CppWriter/ComponentCppWriter/ComponentCppWriterUtils.scala

```diff
@@ -0,0 +1,612 @@
+          Some("The serialization buffer")
+        )
+      )
+      case _ => portParamMap(p.getUnqualifiedName)
+    }
+
+  /** Get a return type of a port instance as an optional C++ type */
+  def getPortReturnType(p: PortInstance): Option[String] =
+    p.getType match {
+      case Some(PortInstance.Type.DefPort(symbol)) =>
+        symbol.node._2.data.returnType match {
+          case Some(typeName) => Some(
+            TypeCppWriter.getName(
+              s,
+              s.a.typeMap(typeName.id),
+              None,
+              PortCppWriter.getPortNamespaces(symbol.getUnqualifiedName)
+            )
+          )
+          case _ => None
+        }
+      case _ => None
+    }
+
+  /** Get a return type of a port as a CppDoc type */
+  def getPortReturnTypeAsCppDocType(p: PortInstance): CppDoc.Type =
+    CppDoc.Type(
+      getPortReturnType(p) match {
+        case Some(tn) => tn
+        case None => "void"
+      }
+    )
+
+  def addReturnKeyword(str: String, p: PortInstance): String =
+    p.getType match {
+      case Some(PortInstance.Type.Serial) => s"return $str"
+      case _ => getPortReturnType(p) match {
+          case Some(_) => s"return $str"
+          case None => str
+        }
+    }
+
+  /** Get the port type as a string */
+  def getPortTypeString(p: PortInstance): String =
+    p match {
+      case _: PortInstance.General => p.getType.get match {
+        case PortInstance.Type.DefPort(_) => "typed"
+        case PortInstance.Type.Serial => "serial"
+      }
+      case _: PortInstance.Special => "special"
+      case _: PortInstance.Internal => "internal"
+    }
+
+  def getPortListTypeString(ports: List[PortInstance]): String =
+    ports match {
+      case Nil => ""
+      case _ => getPortTypeString(ports.head)
+    }
+
+  /** Get the command param type as a string */
+  def getCommandParamString(kind: Command.Param.Kind): String =
+    kind match {
+      case Command.Param.Save => "save"
+      case Command.Param.Set => "set"
+    }
+
+  /** Write an internal port param as a C++ type */
+  def getInternalPortParam(param: Ast.FormalParam): String =
+    TypeCppWriter.getName(
+      s,
+      s.a.typeMap(param.typeName.id),
+      Some("Fw::InternalInterfaceString")
+    )
+
+  /** Write an event param as a C++ type */
+  def getEventParam(param: Ast.FormalParam) =
+    TypeCppWriter.getName(
+      s,
+      s.a.typeMap(param.typeName.id),
+      Some("Fw::LogStringArg")
+    )
+
+  /** Write a channel type as a C++ type */
+  def getChannelType(t: Type): String =
+    TypeCppWriter.getName(s, t, Some("Fw::TlmString"))
+
+  /** Add an optional string separated by two newlines */
+  def addSeparatedString(str: String, strOpt: Option[String]): String = {
+    strOpt match {
+      case Some(s) => s"$str\n\n$s"
+      case None => str
+    }
+  }
+
+  /** Add an optional pre comment separated by two newlines */
+  def addSeparatedPreComment(str: String, commentOpt: Option[String]): String = {
```

Comment:
This function seems generic too.

Reply:
Ok, I moved this function to CppWriterUtils.